cmd/bclient/fileutil/filelist.go

```diff
@@ -0,0 +1,101 @@
+// This package manages a list of files, and their checksums
+
+package fileutil
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"fmt"
+	"github.com/antonholmquist/jason"
+	"io"
+	"os"
+)
+
+// Our underlying data structure
+
+type FileList struct {
+	Root  string
+	Files map[string]map[int64][]byte
```

Comment:
Ugh. I had a flash of insight yesterday morning when this made sense to me,
but it's gone again. I'll leave it until you're back.

On Fri, Jan 22, 2016 at 2:28 PM, Don Brower <notifications@github.com>
wrote:

> In cmd/bclient/fileutil/filelist.go
> <https://github.com/ndlib/bendo/pull/56#discussion_r50579480>:
>
> > +package fileutil
> > +
> > +import (
> > +	"crypto/md5"
> > +	"encoding/base64"
> > +	"fmt"
> > +	"github.com/antonholmquist/jason"
> > +	"io"
> > +	"os"
> > +)
> > +
> > +// Our underlying data structure
> > +
> > +type FileList struct {
> > +	Root  string
> > +	Files map[string]map[int64][]byte
>
> I think this structure can be simplified since the name of the file on the
> remote server doesn't matter. The only thing needed to track is the hashes
> of all the blobs in the item. Thought experiment: The server has a file
> with the contents hello:
>
> slot map:
>     "a" -> 1
>
> blob map:
>     1 (hash 5d41402abc4b2a76b9719d911017c592) -> "hello"
>
> Now we add a file b with the same content. We do not need to add file b,
> because we discover the hash of the content of file b is identical to a
> blob already in the item. So we just make an entry pointing b to the same
> blob:
>
> new slot table
>     "a" -> 1
>     "b" -> 1
>
> new blob table
>     1 (hash 5d41402abc4b2a76b9719d911017c592) -> "hello"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ndlib/bendo/pull/56/files#r50579480>.
>